# <center> Проект: Анализ вакансий из HeadHunter

# Юнит 1. Работа с базой данных из Python

#### Импорт библиотек

In [99]:
import psycopg2
import pandas as pd

# Юнит 2. Введение. Знакомство с данными

In [100]:
connection = psycopg2.connect(
    dbname = ,
    user = ,
    host = ,
    password = ,
    port = 
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [101]:
# текст запроса
query_3_1 = f'''
    select 
        count(id) "Количество вакансий"
    from 
        public.vacancies
    '''

In [102]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\4045908245.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [103]:
# текст запроса
query_3_2 = f'''
    select
        count(id) "Количество работодателей"
    from
        public.employers
    '''

In [104]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\533540996.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


Количество работодателей
0                     23501

3. Посчитайте с помощью запроса количество регионов (таблица areas).

In [105]:
# текст запроса
query_3_3 = f'''
    select
        count(id) "Количество регионов"
    from
        public.areas
    '''

In [106]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\1395976625.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


Количество регионов
0                 1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица industries).

In [107]:
# текст запроса
query_3_4 = f'''
    select
        count(id) "Количество сфер деятельности"
    from
        public.industries
    '''

In [108]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\854197215.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


Количество сфер деятельности
0                           294

***

#### Выводы:
Предварительный анализ вакансий показал, что 23501 работодателей разместили 49197 вакансий разнообразных сфер деятельности (294 сферы деятельности) в 1362 регионах.


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [109]:
# текст запроса
query_4_1 = f'''
    select 
        count(v.id) as cnt,
        a.name as area
    from 
        public.vacancies v
        join public.areas a on a.id = v.area_id
    group by a.name
    order by 1 desc
    limit 10
    '''

In [110]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\620650199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы
5  1698     Екатеринбург
6  1670  Нижний Новгород
7  1415           Казань
8  1301        Краснодар
9  1144           Самара

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [111]:
# текст запроса
query_4_2 = f'''
    select
        count(v.id) "Количество вакансий"
    from
        public.vacancies v
    where salary_to is not null or salary_from is not null
    '''

In [112]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\3746829199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query_4_2, connection)


Количество вакансий
0                24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [113]:
# текст запроса
query_4_3 = f'''
    select
        round(avg(salary_from)) as "lower_bound",
        round(avg(salary_to)) as "upper_bound"
    from
        public.vacancies'''

In [114]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\3595259731.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query_4_3, connection)


lower_bound  upper_bound
0      71065.0     110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.

In [115]:
# текст запроса
query_4_4 = f'''
    select
        count(id) "Количество вакансий",
        schedule "Тип рабочего графика",
        employment "Тип трудоустройства"
    from
        public.vacancies
    group by schedule, employment
    order by 1 desc
    limit 5
    '''

In [116]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\2028972612.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query_4_4, connection)


Количество вакансий Тип рабочего графика  Тип трудоустройства
0                35367          Полный день     Полная занятость
1                 7802     Удаленная работа     Полная занятость
2                 1593        Гибкий график     Полная занятость
3                 1312     Удаленная работа  Частичная занятость
4                  940       Сменный график     Полная занятость

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [117]:
# текст запроса
query_4_5 = f'''
    select
        distinct experience "Опыт работы",
        count(id) "Количество вакансий"
    from
        public.vacancies
    group by experience
    order by 2'''

In [118]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\210463167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query_4_5, connection)


Опыт работы  Количество вакансий
0         Более 6 лет                 1337
1           Нет опыта                 7197
2       От 3 до 6 лет                14511
3  От 1 года до 3 лет                26152

***

#### Выводы:
Детальный анализ вакансий показал следующее:
- Наибольшее количество вакансий было размещено в таких городах, как Москва, Санкт-Петербург, Минск, Новосибирск, Алматы, Екатеринбург. Перечисленные города являются городами - мегаполисами, у которых население свыше 1 миллиона жителей, с большими масштабами урбанизации. Для мегаполисов характерно высокий темп производственно-технологических процессов, хорошо развитая социальная инфраструктура, широкое развитие разнообразных сфер услуг. Чтобы все процессы мегаполиса функционировали необходимо огромное количество персонала.
- Размер заработной платы был указан только 24073 вакансиях, примерно половина от общего количества соискателей не указали желаемую зароботную плату.
- Из представленных вакансий было выявленно что средняя заработная плата по нижней и верхней границам зарплатной вилки составило 71065 у.е и 110537 у.е. соответственно.
- Самый популярный график работы это полный день с полной занятостью, затем идет удаленная работа с полной  занятостью.
- Наибольшее количество соискателей с опытом работы от 1 года до 3 лет, наименьшее количество соискателей с опытом работы более 6 лет. Можно предположить что чем старше человек, тем он менее склонен к смене места работы.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [119]:
# текст запроса
query_5_1 = f'''
    (select
        e.name "Работодатель",
        count(v.id) "Количество вакансий"
    from
        public.employers e
        join public.vacancies v on e.id = v.employer_id
    group by e.name
    order by 2 desc
    limit 1)
    
    union all
    
    (select
        e.name "Работодатель",
        count(v.id) "Количество вакансий"
    from
        public.employers e
        join public.vacancies v on e.id = v.employer_id
    group by e.name
    order by 2 desc
    offset 4
    limit 1)
    
    '''

In [120]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\1147638897.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query_5_1, connection)


Работодатель  Количество вакансий
0         Яндекс                 1933
1  Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.

In [121]:
# текст запроса
query_5_2 = f'''
    select
        a.name "Регион",
        count(e.id) "Количество работодателей",
        count(v.id) "Количество вакансий"
    from 
        public.areas a
        full join public.employers e on e.area = a.id
        full join public.vacancies v on v.area_id = a.id
    group by a.name
    having count(v.id) = 0
    order by 2 desc
'''


In [122]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\2306256317.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query_5_2, connection)


Регион  Количество работодателей  \
0                          Россия                       410   
1                       Казахстан                       207   
2              Московская область                        75   
3              Краснодарский край                        19   
4              Ростовская область                        18   
..                            ...                       ...   
588                      Струнино                         0   
589    Славгород (Алтайский край)                         0   
590                     Сорочинск                         0   
591  Ленинский (Тульская область)                         0   
592                         Тавда                         0   

     Количество вакансий  
0                      0  
1                      0  
2                      0  
3                      0  
4                      0  
..                   ...  
588                    0  
589                    0  
590                    0  
591                    0  
592                    0  

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.

In [123]:
# текст запроса
query_5_3 = f'''
    select 
        e.name "Работодатель",
        count(distinct v.area_id) "Количество регионов" 
    from 
        public.employers e
        join public.vacancies v on e.id = v.employer_id
    group by 1
    order by 2 desc
        '''

In [124]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\133400711.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


Работодатель  Количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
14761                  UniSol                    1
14762            UNISTORY LLC                    1
14763                   UNIT6                    1
14764     United Distribution                    1
14765              UNITEL LLC                    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [125]:
# текст запроса
query_5_4 = f'''
    select
        count(*) "Количество работодателей"
    from
        public.employers_industries ei
        full join public.employers e on e.id = ei.employer_id
    where industry_id is null
     
        '''

In [126]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\1045119256.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [127]:
# текст запроса
query_5_5 = f'''
    select
        e.name "Компания"
    from
        public.employers e
        join public.employers_industries ei on ei.employer_id = e.id
    group by 1
    having count(ei.industry_id) = 4
    order by 1
    offset 2
    limit 1      
    '''


In [128]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\4025176131.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


Компания
0     2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.

In [79]:
# текст запроса
query_5_6 = f'''
    select
        count(ei.employer_id) "Количество работодателей"
    from
        public.employers_industries ei 
        full join public.industries i on i.id = ei.industry_id
    where i.name = 'Разработка программного обеспечения' 
    '''

In [80]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\2713946032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


Количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [81]:
# код для получения списка городов-милионников

# автоматическое перекодирование ссылок
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

url

# извлечение таблицы из страницы Википедии
table = pd.read_html(url)
df = pd.DataFrame(table[1])
df=df['Город'].replace('Воронеж[a]', 'Воронеж') 
cities=tuple(df) # представление данных в виде кортежа
cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [82]:
# текст запроса
query_5_7 = f'''
    select
        a.name "Города-миллионики",
        count(v.id) "Количество вакансий"
    from
        public.vacancies v
        full join public.areas a on v.area_id = a.id
        full join public.employers e on v.employer_id = e.id
    where a.name in ('Москва','Санкт-Петербург', 'Новосибирск','Екатеринбург', 'Казань','Нижний Новгород', 'Челябинск','Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону','Омск', 'Краснодар','Воронеж', 'Пермь', 'Волгоград') and e.name = 'Яндекс'
    group by 1
    
    union all
    
    select
        'Total',
        count(v.id)
    from
        public.vacancies v
        full join public.areas a on v.area_id = a.id
        full join public.employers e on v.employer_id = e.id
    where a.name in ('Москва','Санкт-Петербург', 'Новосибирск','Екатеринбург', 'Казань','Нижний Новгород', 'Челябинск','Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону','Омск', 'Краснодар','Воронеж', 'Пермь', 'Волгоград') and e.name = 'Яндекс'
    order by 2 

    
'''

In [83]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\2595257991.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


Города-миллионики  Количество вакансий
0               Омск                   21
1          Челябинск                   22
2         Красноярск                   23
3          Волгоград                   24
4              Пермь                   25
5             Казань                   25
6     Ростов-на-Дону                   25
7                Уфа                   26
8             Самара                   26
9          Краснодар                   30
10           Воронеж                   32
11       Новосибирск                   35
12   Нижний Новгород                   36
13      Екатеринбург                   39
14   Санкт-Петербург                   42
15            Москва                   54
16             Total                  485

***

#### Выводы:
Анализ работодателй показал:
- Наиболее популярный работодатель по количеству вакансий (1933 шт) - "Яндекс". Это было наверно ожидаемым, так как Яндекс известен не только поисковой системой, но и различными сервисами, такими как Яндекс.Такси, Яндекс.Маркет, Яндекс.Еда(служба доставки продуктов), Яндекс.Лавка (экспресс-доставка), Яндекс.Драйв (каршеринг) и т.п. На любой из этих сервисов всегда будут рабочие места.
- Яндекс разместил 485 вакансий в городах миллиониках.
- Есть много работодателей среди различных регионов, при этом у них нет открытых вакансий.
- По количеству регионов размещения вакансий лидирует Яндекс. Возможно это связано с тем, что сферы деятельности Яндекса превалирует IT область, что позволяет компании нанимать сотрудников на удаленную работу. Удаленная работа, как известно, - это возможность работать из любой точки мира   
Затем идет "Ростелеком" -  российский провайдер услуг широкополосного доступа в Интернет, интерактивного телевидения, сотовой связи, местной и дальней телефонной связи. Занимает лидирующие позиции на российском рынке высокоскоростного доступа в интернет, платного ТВ, хранения и обработки данных, а также кибербезопасности. В структуре Ростелекома 79 филиалов, расположеннных по всей стране.
- 8419 работодателей не указали в размещаемых ими вакансиях сферы деятельности.
- 1136 работодателей имеют 4 сферы деятельности
- 3553 работодателей занимаются разработкой программного обеспечения

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 

In [84]:
# текст запроса
query_6_1 = f'''
    select
        count(name) "Вакансии, связанные с данными"
    from
        public.vacancies
    where lower(name) like '%data%' or lower(name) like '%данн%' or upper(name) like '%data%' or upper(name) like '%данн%'
    '''

In [85]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\1281472032.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query_6_1, connection)


Вакансии, связанные с данными
0                           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.

In [86]:
# текст запроса
query_6_2 = f'''
    select
        count(v.id) "Количество вакансий DS-junior"
    from
        public.vacancies v
    where (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        lower(v.name) like '%machine learning%' or
        lower(v.name) like '%машинн%обучен%' or 
        (v.name like '%ML%' and v.name not like '%HTML%'))
        and 
        (lower(v.name) like '%junior%' or 
        v.experience like '%Нет опыта%' or 
        v.employment like '%Стажировка%')
    
    '''

In [87]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\118162426.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query_6_2, connection)


Количество вакансий DS-junior
0                             51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [88]:
# текст запроса
query_6_3 = f'''
    select
        count(v.id) "Количество вакансий с ключевыми навыками SQL или postgres"
    from
        public.vacancies v
    where (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        lower(v.name) like '%machine learning%' or
        lower(v.name) like '%машинн%обучен%' or 
        (v.name like '%ML%' and v.name not like '%HTML%'))
        and 
        (key_skills like '%SQL%' or
        key_skills like '%postgres%')
    '''

In [89]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\3310962512.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query_6_3, connection)


Количество вакансий с ключевыми н
0                                201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [129]:
# текст запроса
query_6_4 = f'''
    select
        count(v.id) "Количество вакансий с ключевым навыком Python"
    from
        public.vacancies v
    where (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        lower(v.name) like '%machine learning%' or
        lower(v.name) like '%машинн%обучен%' or 
        (v.name like '%ML%' and v.name not like '%HTML%'))
        and 
        (key_skills like '%Python%')
    '''

In [130]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\614304167.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query_6_4, connection)


Количество вакансий с ключевым на
0                                351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [92]:
# текст запроса
query_6_5 = f'''
    select
        round(avg(length(key_skills)-
        length(replace(key_skills, chr(9), ''))+1),2) "Количество ключевых нвыков для DS"
    from
       public.vacancies v
    where (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        lower(v.name) like '%machine learning%' or
        lower(v.name) like '%машинн%обучен%' or 
        (v.name like '%ML%' and v.name not like '%HTML%')) 
'''

In [93]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\4089566367.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query_6_5, connection)


Количество ключевых нвыков для DS
0                               6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [94]:
# текст запроса
query_6_6 = f'''
    select
        experience "Опыт работы",
        round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to))) "Средняя ЗП"
    from 
        public.vacancies v
    where (lower(v.name) like '%data scientist%' or 
        lower(v.name) like '%data science%' or 
        lower(v.name) like '%исследователь данных%' or 
        lower(v.name) like '%machine learning%' or
        lower(v.name) like '%машинн%обучен%' or 
        (v.name like '%ML%' and v.name not like '%HTML%'))  
    group by 1 
        
    '''

In [95]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

C:\Users\Елена\AppData\Local\Temp\ipykernel_9096\2297777544.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query_6_6, connection)


Опыт работы  Средняя ЗП
0         Более 6 лет         NaN
1           Нет опыта     74643.0
2  От 1 года до 3 лет    139675.0
3       От 3 до 6 лет    243115.0

***

#### Выводы:
С помощью предметного анализа было выявлено:
- 1771 вакансий, связанных с данными
- при этом 51 вакансия размещена на позицию DS-junior
- в 201 вакансиях упоминается необходимые навыки такие как SQL или postgres
- в 351 вакансиях - навык владения языком программирования Python
- в вакансиях на позицию DS в среднем упоминается больше 6 ключевых навыков
-  DS с опытом работы от 3 до 6 лет может расчитывать в среднем на ЗП в размере 243115 у.е.

In [96]:
connection.close()

# Общий вывод по проекту

1. В выборке представлены 49197 вакансий от 23501 работодателей в 1362 регионах по 294 сферам деятельсности.
2. Наибольшее количество вакансий было размещено в городах-мегаполисах с хорошо развитыми промышленностью и инфраструктурой.
3. Средняя зароботная плата составляет от 71065 у.е. до 110537 у.е.
4. По количеству размещенных вакансий лидирует "Яндекс"
5. 36% работодателей не указали в своих вакансиях сферу деятельности
6. 10% работадателей указали 4 и более сферы деятельности
7. 15% работодателей занимаются разработкой программного обеспечения
8. Из 1771 вакансий, связанных с данными только в 201 вакансиях требуется навык владения SQL или postgres, в 351 вакансиях - Pythonю
9. В вакансиях для DS в среднем указывают более 6 ключевых навыков. 
10. DS с опытом работы от 3 до 6 лет будет получать в среднем на ЗП в 243115 у.е.




